# CLUES2 Companion: Pipelines for estimating and visualizing selection coefficients as well as dating the onset of selective sweeps for multilocus sites 

---

**Overview**

**CLUES2Companion.sh** is a comprehensive three-phase pipeline that includes:

- **Phase 1**:  
  - Converts VCF → `.haps` & `.sample`  
  - Prepares input (masking, flipping, filtering)  
  - Runs Relate (two passes to `.anc`, `.mut`, `.coal`)  
  - Extracts SNPs, polarized derived allele per SNP, and derived allele frequency table  

- **Phase 2**:  
  - Applies importance sampling of branch lengths to generate Newick trees based on multilocus sites  
  - Runs **RelateToCLUES.py** on multilocus sites within a target region `*_times.txt`  
  - Runs **CLUES2** (`inference.py`) on multilocus sites within a target region `_inference.txt` (+ optional `_CI.txt`)  
  - Merges information on multiple SNPs into a *.tsv file  
  - Integrative plotting (showing error bars, −log₁₀(p), derived allele frequencies, etc.)  

- **Phase-3**:  
  - Scans different time bins to date the onset of selection on SNP of interest (in generations & years)  
  - Performs optional bootstraps with user-defined settings to calculate confidence intervals around the estimated onset of selection   
  - Reporting of date of onset and corresponding confidence intervals in `*.json` file   

Outputs for Phase 1, Phase 2, and Phase 3, including log files, are automatically saved to output folders generated by CLUES2 Companion (e.g., Output2Companion/phase1/<PREFIX>_chr<CHR>).

**[Important]:** Users must keep the following folders exactly where they are

Moving or renaming **any** of these folders – or this script itself –
will interfere with the functioning of Phase 1, Phase 2 and Phase 3 algorithms.
---

## Table of Contents

- [Installation](#Installation)  
- [Dependencies](#dependencies)  
- [Input files & folder layout](#inputs)  
- [Quick start](#quick-start)  
- [Phase 1 (details)](#Phase 1)  
- [Phase 2 (details)](#Phase 2)  
- [Phase 3 (details)](#Phase 3)  
- [Tips & troubleshooting](#tips)  
- [License](#license)  
- [Contact](#contact)  

---

<a name="Installation"></a>
## Installation

```bash
git clone https://github.com/alisi1989/CLUES2_Companion.git
cd CLUES2_Companion
chmod +x CLUES2Companion.sh
```
Ensure your working dir contains the following sub-folders:

```
Relate/     contains Relate binary files  
CLUES2/     contains CLUES2 scripts  
required_files/  
  ├ ancestor/homo_sapiens_ancestor_chrN.fa  
  ├ mask/PilotMask_chrN.fasta  
  └ map/genetic_map_chrN.txt  
```

<a name="dependencies"></a>

## Dependencies

Bash ≥ 4.0 (for mapfile) \
Python 3.8+ \
Relate v1.2.2 \
CLUES2 (master branch) \
GNU parallel (optional) \
Python packages: numpy, pandas, matplotlib, adjustText

```
Example of installation of Python packages

pip3 install cyvcf2 numpy pandas matplotlib adjustText
```

<a name="inputs"></a>

Input file & folder layout tree

```
CLUES2 Companion/
├── CLUES2Companion.sh         # main driver
├── Relate/                    # Relate bin/, scripts/
├── CLUES2/                    # CLUES v2 scripts
├── required_files/            # reference data
│   ├ ancestor/…
│   ├ mask/…
│   └ map/…
└── example/
    ├ Finnish_chr2.vcf.gz      # phased and indexed vcf
    └ Finnish.poplabels        # sampleID, population, group, SEX
```
Users must use fully phased vcf and indexed files (e.g., *.vcf.gz and *.tbi files). 

Furthermore, the vcf file must contain only one population and one chromosome (e.g., example/Finnish_chr2.vcf.gz)

In addition, the *.poplabels file must contain four columns (namely, sampleID, population, group, SEX)

Example of *.poplabels file for diploid organisms:
```
sample population group sex
UNR1 FIN EUR NA
UNR2 FIN EUR NA
UNR3 FIN EUR NA
UNR4 FIN EUR NA
```
For more information please refer to: `https://myersgroup.github.io/relate/input_data.html#Prepare`

---

<a name="quick-start"></a>

## Quick start:
```
./CLUES2Companion.sh
```

Menu prompts: choose Phase 1, Phase 2 or Phase 3.


<a name="phase-1"></a>

## Phase 1 – Run Relate to create input files for CLUES2 

1 - Convert vcf to `*.haps`, `*.sample` \
2 - Apply PrepareInputFiles.sh in Relate (to mask, flip, and filter SNPs) \
3 - Run Relate mode All (with a user-specified Ne to generate the `*.anc` and `*.mut` files \
4 - Apply EstimatePopulationSize.sh to generate the `*.coal` file \
5 - Re-estimate branch lengths using the `*.coal` to generated updated `*.anc` and `*.mut` files \
6 - Use the cyvcf2 package to extract, SNPs and corresponding positions within a user-specified target region \
7 - Polarize derived alleles and compute derived allele frequency \

## Example of usage of Phase 1

```
./CLUES2Companion.sh
******  CLUES2 Companion – please cite CLUES2 and CLUES2Companion  ******
Choose phase to run
  1) Phase 1  : Apply Relate (*.mut, *.anc, *.coal files along with derived allele frequency) 
  2) Phase 2  : Apply Relate (BranchLengths) and CLUES2 (RelateToCLUES.py and inference.py) 
  3) Phase 3  : Date onset of selective sweeps of target SNP(s)
Enter option (1/2/3):
```
Here, users have to make their choice. That is to say, they will have to enter either 1, 2, or 3.

`then`

```
Enter option (1/2/3): 1   


          ╔═════════════════════════════════════════════════════════╗
          ║        🚀  PHASE 1: RELATE & SNP EXTRACTION  🚀        ║
          ║   Please read the manual carefully before proceeding    ║
          ╚═════════════════════════════════════════════════════════╝

Choose the chromosome to analyze (e.g. 2, 17, X): 2
Prefix of phased vcf/bcf file (e.g. example/Finnish without _chrN.vcf.gz): example/Finnish
Path to population‑labels file (*.poplabels)(e.g. example/Finnish.poplabels): example/Finnish.poplabels
Start bp of target region: 135839626
End bp of target region: 135876443
Prefix of output name (e.g. Finnish): Finnish
```
### Explanation of Phase 1: required inputs and on-screen feedback

`1 - Select the input data:`

| Prompt                                                          | What to type                   |
| --------------------------------------------------------------- | ------------------------------ |
| `Chromosome to analyze (e.g. 2, 17, X):`                        | chromosome number (e.g., 2)    |
| `Prefix of phased vcf/bcf (Finnish without _chrN.vcf.gz):`      | e.g. `example/Finnish`         |


`2 - Define the target region"`

| Prompt                       | Description                                       |
| ---------------------------- | ------------------------------------------------- |
| `Start bp of target region:` | genomic start coordinate of the window to analyse |
| `End bp of target region:`   | genomic end coordinate of the window to analyse   |


All SNPs between these positions are extracted from the fully phased vcf.

`3 - Choose an output prefix:`

| Prompt                                  | Description                                                                                     |
| --------------------------------------- | ----------------------------------------------------------------------------------------------- |
| `Prefix of output name:`                | e.g. `example/Finnish` **Important: You must use the exact same prefix in Phase 2 and Phase 3** |


### What you will see on the screen:

Every major step prints an INFO line that indicates where the resulting files are saved (e.g., output_C2Companion/phase1/)

While a step is running, a progress bar will move across the terminal. If something fails, you get a [WARN] or [ERROR] message; otherwise the step ends with the word `done!`.


<a name="phase-2"></a>

## Phase-2 – Selection coefficient inference

1 - Apply SampleBranchLengths.sh (Relate) to sample ancestral recombination graphs (ARGs) to generate Newick trees \
2 - Apply RelateToCLUES.py (CLUES2) to generate *_times.txt \
3 - Apply inference.py (CLUES2) to estimate selection coefficients and confidence intervals \
4 - Merge summary statistics, rsID, genomic coordinates, and derived allele frequency into *.tsv file \
5 - Generate tabular and graphical outputs \

---

### Example of usage for Phase 2

```
******  CLUES2Companion – please cite CLUES2 and CLUES2Companion  ******
Choose phase to run
  1) Phase 1  : Apply Relate (*.mut, *.anc, *.coal files along with derived allele frequency) 
  2) Phase 2  : Apply Relate (BranchLengths) and CLUES2 (RelateToCLUES.py and inference.py) 
  3) Phase 3  : Date onset of selective sweeps of target SNP(s)
Enter option (1/2/3): 2

```

`then`

```
          ╔═════════════════════════════════════════════════════════╗
          ║        🧬  PHASE-2 – (requires Phase‑1 outputs)  🧬     ║
          ║   Please read the manual carefully before proceeding    ║
          ╚═════════════════════════════════════════════════════════╝

Choose the chromosome to analyze (e.g. 2, 17, X): 2
Enter population prefix used in Phase‑1 (e.g., Finnish_MCM6): Finnish
Phase 1 auto-detect directory [ENTER = /~/CLUES2Companion2/output_C2Companion/phase1/Finnish_MCM6_chr2] or provide a different folder with phase1 outputs :

→ Using frequency file: ~/CLUES2Companion2/output_C2Companion/phase1/FIN_MCM6_chr2/Finnish_MCM6_Frequency_chr2_135839626_135876443.txt
→ Using SNPs file: ~/CLUES2Companion2/output_C2Companion/phase1/FIN_MCM6_chr2/Finnish_MCM6_SNPs_chr2_135839626_135876443.txt

tCutoff (e.g. 1000): 500
df (e.g. 600): 600
importance sampling of branch lengths: 200
AncientSamps file (optional, ENTER to skip): 
AncientHaps  file (optional, ENTER to skip): 
Disable allele trajectory? (y/N): y
Confidence interval (e.g. 0.95) [ENTER to skip]: 0.95
TimeBins (a list of epoch breakpoints; optional, ENTER to skip):
```

## Explanation of above command lines

### Script prompts:
Select the chromosome to analyze:
**Chromosome to analyze (e.g., 2, 17, X):**
If you run Phase 1 on several chromosomes, the pipeline automatically searches for the Phase 1 output that matches the chromosome that you enter above.

**Enter the population prefix:**
Provide the same prefix you used in Phase 1 (e.g. Finnish). The Phase 2 script now knows the exact folder structure (e.g., output_C2Companion/phase1/Finnish_chr2/).

**Confirm detected Phase 1 folders:**
The script shows the path to the location of the following file: *.anc, *.mut, *.coal, SNP list (`*.txt`), and frequency table (`*.txt`).
If the path is correct, press Enter to accept.
If users have manually moved Phase 1 folder (which is not recommended), type the path to the location of the new directory and press Enter.

### Mandatory CLUES2 parameters (please see CLUES2 and Relate manuals):

| Prompt                                    | Description                                          | Reference              |
| ----------------------------------------- | ---------------------------------------------------- | ---------------------- |
| `tCutoff (e.g. 1000):`                    | The maximum time (in generations)                    | CLUES2 `--tCutoff`    |
| `df (e.g. 600):`                          | number of allele frequency bins                      | CLUES2 `--df`         |
| `Importance sampling of branch lengths:`  | number of trees to sample from *Relate*              | Relate `--num_samples` |

None of the above prompts can be left blank.

### Optional files (please see CLUES2 and Relate manuals):

`AncientSamps — table of sample ages`
`AncientHaps — ancient haplotypes`
`Press Enter to skip either file`

Please see the CLUES2 manual for examples of formatting

### Optional parameters to include (please see CLUES2 and Relate manuals):

| prompt                                  | effect                                                              |
| --------------------------------------- | ------------------------------------------------------------------- |
| `Disable allele trajectory? (y/N):`     | answer **y** to *skip* posterior trajectories speeds up inference   |
| `Confidence intervals (e.g. 0.95):`     | compute CI for the selection coefficient (MLE); **Enter** = no CI   |
| `TimeBins (e.g., 200 300):`             | split 0-*tCutoff* into custom intervals; *Enter* = single epoch     |


The example TimeBins 200 300 for a tCutoff 500 yields three epochs: 0-200, 200-300, 300-500.

### Internal housekeeping:

NOTE: All output files will be saved to `~/Output_C2Companion/phase2/{prefix}_chr{N}/`. In this example output files are saved in `~/Output_C2Companion/phase2/FIN_MCM6_original_chr2/)`

### The resulting plot will contain:

SNPs analyzed in Phase 2 \
Confidence intervals \ 
Color intensity bar `[−log₁₀(p)]` \ 
Astericks indicating significance above each SNP (* (0.05), ** (0.01), *** (0.001)) 


### A *.tsv file containing SNP information, including related statistics calculated by CLUES2. 

**Example of `*.tsv file`**
```
rsID	POS	der_freq	logLR	-log10(p)	Epoch1_start	Epoch1_end	SelectionMLE1	95%_lower	95%_upper
rs55809728	135842606	0.0909	3.5508	2.11	0	536	0.09981	0.05336	0.14626
rs3754686	135845706	0.6515	5.3420	2.97	0	536	0.02312	0.00405	0.04218
rs4988243	135850133	0.1263	2.0500	1.37	0	536	0.09984	-0.01286	0.21254
rs4954490	135850661	0.6515	5.5070	3.04	0	536	0.02308	0.00369	0.04247
rs4988235	135851076	0.5909	17.2338	8.36	0	536	0.09986	0.08678	0.11294
rs4954493	135852405	0.6515	5.7506	3.16	0	536	0.02420	0.00069	0.04772
rs4988226	135853028	0.6515	5.3420	2.97	0	536	0.02312	0.00405	0.04218
rs309178	135854054	0.6515	5.5244	3.05	0	536	0.02288	0.00368	0.04208
rs309179	135856210	0.6515	5.4909	3.04	0	536	0.02248	0.00139	0.04356
```
[INFO] In addition to the final outputs, users will find three additional folders (`{prefix_inference_chr{N}/`, `prefix_times_chr{N}/`, `prefix_trees_chr{N}/` in which the output from inference.py, RelateToClues.py, and Relate are stored, respectively)

**FuAdditional reading**

CLUES2 manual: https://github.com/avaughn271/CLUES2#command-line-arguments \

Relate SampleBranchLengths: https://myersgroup.github.io/relate/modules.html#SampleBranchLengths \

---

<a name="phase-3"></a>

Phase-3 – Dating selective sweeps

Initial scan over fixed windows (0–2000 gen) → onset estimate
Save initial JSON:

`{ "rsID": "...", "onset_gen": 100, "onset_years": 2800, … }`


Optional bootstrap: user-defined epochs, window size, replicates
Compute per-replicate onsets → summary statistics (median, 95% CI)
Save JSON with CI:
rsID_Boostraps_onset_Dating+CI.json

<a name="tips"></a>

Tips & troubleshooting

Logging: check output_C2Companion/log/ for debug
Missing deps: verify RelateFileFormats, PrepareInputFiles.sh, Python packages
Spinner issues: ensure bash -euo pipefail is supported
Plot errors: install adjustText for label adjustment
Bootstrap: skip by answering “N” when prompted
<a name="license"></a>

License

MIT License

<a name="contact"></a>

Contact

Alessandro Lisi alisi@usc.edu (alisi1989)
Michael C. Campbell mc44680@usc.edu (mc44680)
