# CLUES2 Companion: Pipelines for estimating and visualizing selection coefficients as well as dating the onset of selective sweeps for multilocus sites 

---

**Overview**

**CLUES2Companion.sh** is a comprehensive three-phase pipeline that includes:

- **Phase 1**:  
  - Converts VCF â†’ `.haps` & `.sample`  
  - Prepares input (masking, flipping, filtering)  
  - Runs Relate (two passes to `.anc`, `.mut`, `.coal`)  
  - Extracts SNPs, polarized derived allele per SNP, and derived allele frequency table  

- **Phase 2**:  
  - Applies importance sampling of branch lengths to generate Newick trees based on multilocus sites  
  - Runs **RelateToCLUES.py** on multilocus sites within a target region `*_times.txt`  
  - Runs **CLUES2** (`inference.py`) on multilocus sites within a target region `_inference.txt` (+ optional `_CI.txt`)  
  - Merges information on multiple SNPs into a *.tsv file  
  - Integrative plotting (showing error bars, âˆ’logâ‚â‚€(p), derived allele frequencies, etc.)  

- **Phase-3**:  
  - Scans different time bins to date the onset of selection on SNP of interest (in generations & years)  
  - Performs optional bootstraps with user-defined settings to calculate confidence intervals around the estimated onset of selection   
  - Reporting of date of onset and corresponding confidence intervals in `*.json` file   

Outputs for Phase 1, Phase 2, and Phase 3, including log files, are automatically saved to output folders generated by CLUES2 Companion (e.g., Output2Companion/phase1/<PREFIX>_chr<CHR>).

**[Important]:** Users must keep the following folders exactly where they are

Moving or renaming **any** of these folders â€“ or this script itself â€“
will interfere with the functioning of Phase 1, Phase 2 and Phase 3 algorithms.
---

## Table of Contents

- [Installation](#Installation)  
- [Dependencies](#dependencies)  
- [Input files & folder layout](#inputs)  
- [Quick start](#quick-start)  
- [Phase 1 (details)](#Phase1)  
- [Phase 2 (details)](#Phase2)  
- [Phase 3 (details)](#Phase3)  
- [Tips & troubleshooting](#tips)  
- [License](#license)  
- [Contact](#contact)  

---

<a name="Installation"></a>
## Installation

```bash
git clone https://github.com/alisi1989/CLUES2_Companion.git
cd CLUES2_Companion
chmod +x CLUES2Companion.sh
```
Ensure your working dir contains the following sub-folders:

```
Relate/     contains Relate binary files  
CLUES2/     contains CLUES2 scripts  
required_files/  
  â”œ ancestor/homo_sapiens_ancestor_chrN.fa  
  â”œ mask/PilotMask_chrN.fasta  
  â”” map/genetic_map_chrN.txt  
```

<a name="dependencies"></a>

## Dependencies

Bash â‰¥ 4.0 (for mapfile) \
Python 3.8+ \
Relate v1.2.2 \
CLUES2 (master branch) \
GNU parallel (optional) \
Python packages: numpy, pandas, matplotlib, adjustText

```
Example of installation of Python packages

pip3 install cyvcf2 numpy pandas matplotlib adjustText
```

<a name="inputs"></a>

Input file & folder layout tree

```
CLUES2 Companion/
â”œâ”€â”€ CLUES2Companion.sh         # main driver
â”œâ”€â”€ Relate/                    # Relate bin/, scripts/
â”œâ”€â”€ CLUES2/                    # CLUES v2 scripts
â”œâ”€â”€ required_files/            # reference data
â”‚   â”œ ancestor/â€¦
â”‚   â”œ mask/â€¦
â”‚   â”” map/â€¦
â””â”€â”€ example/
    â”œ Finnish_chr2.vcf.gz      # phased and indexed vcf
    â”” Finnish.poplabels        # sampleID, population, group, SEX
```
Users must use fully phased vcf and indexed files (e.g., *.vcf.gz and *.tbi files). 

Furthermore, the vcf file must contain only one population and one chromosome (e.g., example/Finnish_chr2.vcf.gz)

In addition, the *.poplabels file must contain four columns (namely, sampleID, population, group, SEX)

Example of *.poplabels file for diploid organisms:
```
sample population group sex
UNR1 FIN EUR NA
UNR2 FIN EUR NA
UNR3 FIN EUR NA
UNR4 FIN EUR NA
```
For more information please refer to: `https://myersgroup.github.io/relate/input_data.html#Prepare`

---

<a name="quick-start"></a>

## Quick start:
```
./CLUES2Companion.sh
```

Menu prompts: choose Phase 1, Phase 2 or Phase 3.


<a name="Phase1"></a>

## Phase 1 â€“ Run Relate to create input files for CLUES2 

1 - Convert vcf to `*.haps`, `*.sample` \
2 - Apply PrepareInputFiles.sh in Relate (to mask, flip, and filter SNPs) \
3 - Run Relate mode All (with a user-specified Ne to generate the `*.anc` and `*.mut` files \
4 - Apply EstimatePopulationSize.sh to generate the `*.coal` file \
5 - Re-estimate branch lengths using the `*.coal` to generated updated `*.anc` and `*.mut` files \
6 - Use the cyvcf2 package to extract, SNPs and corresponding positions within a user-specified target region \
7 - Polarize derived alleles and compute derived allele frequency \

## Example of usage of Phase 1

```
./CLUES2Companion.sh
******  CLUES2 Companion â€“ please cite CLUES2 and CLUES2Companion  ******
Choose phase to run
  1) Phase 1  : Apply Relate (*.mut, *.anc, *.coal files along with derived allele frequency) 
  2) Phase 2  : Apply Relate (BranchLengths) and CLUES2 (RelateToCLUES.py and inference.py) 
  3) Phase 3  : Date onset of selective sweeps of target SNP(s)
Enter option (1/2/3):
```
Here, users have to make their choice. That is to say, they will have to enter either 1, 2, or 3.

`then`

```
Enter option (1/2/3): 1   


          â•”â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•—
          â•‘        ğŸš€  PHASE 1: RELATE & SNP EXTRACTION  ğŸš€        â•‘
          â•‘   Please read the manual carefully before proceeding    â•‘
          â•šâ•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•

Choose the chromosome to analyze (e.g. 2, 17, X): 2
Prefix of phased vcf/bcf file (e.g. example/Finnish without _chrN.vcf.gz): example/Finnish
Path to populationâ€‘labels file (*.poplabels)(e.g. example/Finnish.poplabels): example/Finnish.poplabels
Start bp of target region: 135839626
End bp of target region: 135876443
Prefix of output name (e.g. Finnish): Finnish
```
### Explanation of Phase 1: required inputs and on-screen feedback

`1 - Select the input data:`

| Prompt                                                          | What to type                   |
| --------------------------------------------------------------- | ------------------------------ |
| `Chromosome to analyze (e.g. 2, 17, X):`                        | chromosome number (e.g., 2)    |
| `Prefix of phased vcf/bcf (Finnish without _chrN.vcf.gz):`      | e.g. `example/Finnish`         |


`2 - Define the target region"`

| Prompt                       | Description                                       |
| ---------------------------- | ------------------------------------------------- |
| `Start bp of target region:` | genomic start coordinate of the window to analyse |
| `End bp of target region:`   | genomic end coordinate of the window to analyse   |


All SNPs between these positions are extracted from the fully phased vcf.

`3 - Choose an output prefix:`

| Prompt                                  | Description                                                                                     |
| --------------------------------------- | ----------------------------------------------------------------------------------------------- |
| `Prefix of output name:`                | e.g. `example/Finnish` **Important: You must use the exact same prefix in Phase 2 and Phase 3** |


### What you will see on the screen:

Every major step prints an INFO line that indicates where the resulting files are saved (e.g., output_C2Companion/phase1/)

While a step is running, a progress bar will move across the terminal. If something fails, you get a [WARN] or [ERROR] message; otherwise the step ends with the word `done!`.


<a name="Phase2"></a>

## Phase-2 â€“ Selection coefficient inference

1 - Apply SampleBranchLengths.sh (Relate) to sample ancestral recombination graphs (ARGs) to generate Newick trees \
2 - Apply RelateToCLUES.py (CLUES2) to generate *_times.txt \
3 - Apply inference.py (CLUES2) to estimate selection coefficients and confidence intervals \
4 - Merge summary statistics, rsID, genomic coordinates, and derived allele frequency into *.tsv file \
5 - Generate tabular and graphical outputs \

---

### Example of usage for Phase 2

```
******  CLUES2Companion â€“ please cite CLUES2 and CLUES2Companion  ******
Choose phase to run
  1) Phase 1  : Apply Relate (*.mut, *.anc, *.coal files along with derived allele frequency) 
  2) Phase 2  : Apply Relate (BranchLengths) and CLUES2 (RelateToCLUES.py and inference.py) 
  3) Phase 3  : Date onset of selective sweeps of target SNP(s)
Enter option (1/2/3): 2

```

`then`

```
          â•”â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•—
          â•‘        ğŸ§¬  PHASE-2 â€“ (requires Phaseâ€‘1 outputs)  ğŸ§¬     â•‘
          â•‘   Please read the manual carefully before proceeding    â•‘
          â•šâ•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•

Choose the chromosome to analyze (e.g. 2, 17, X): 2
Enter population prefix used in Phaseâ€‘1 (e.g., Finnish_MCM6): Finnish
Phase 1 auto-detect directory [ENTER = /~/CLUES2Companion2/output_C2Companion/phase1/Finnish_MCM6_chr2] or provide a different folder with phase1 outputs :

â†’ Using frequency file: ~/CLUES2Companion2/output_C2Companion/phase1/FIN_MCM6_chr2/Finnish_MCM6_Frequency_chr2_135839626_135876443.txt
â†’ Using SNPs file: ~/CLUES2Companion2/output_C2Companion/phase1/FIN_MCM6_chr2/Finnish_MCM6_SNPs_chr2_135839626_135876443.txt

tCutoff (e.g. 1000): 500
df (e.g. 600): 600
importance sampling of branch lengths: 200
AncientSamps file (optional, ENTER to skip): 
AncientHaps  file (optional, ENTER to skip): 
Disable allele trajectory? (y/N): y
Confidence interval (e.g. 0.95) [ENTER to skip]: 0.95
TimeBins (a list of epoch breakpoints; optional, ENTER to skip):
```

## Explanation of above command lines

### Script prompts:
Select the chromosome to analyze:
**Chromosome to analyze (e.g., 2, 17, X):**
If you run Phase 1 on several chromosomes, the pipeline automatically searches for the Phase 1 output that matches the chromosome that you enter above.

**Enter the population prefix:**
Provide the same prefix you used in Phase 1 (e.g. Finnish). The Phase 2 script now knows the exact folder structure (e.g., output_C2Companion/phase1/Finnish_chr2/).

**Confirm detected Phase 1 folders:**
The script shows the path to the location of the following file: *.anc, *.mut, *.coal, SNP list (`*.txt`), and frequency table (`*.txt`).
If the path is correct, press Enter to accept.
If users have manually moved Phase 1 folder (which is not recommended), type the path to the location of the new directory and press Enter.

### Mandatory CLUES2 parameters (please see CLUES2 and Relate manuals):

| Prompt                                    | Description                                          | Reference              |
| ----------------------------------------- | ---------------------------------------------------- | ---------------------- |
| `tCutoff (e.g. 1000):`                    | The maximum time (in generations)                    | CLUES2 `--tCutoff`    |
| `df (e.g. 600):`                          | number of allele frequency bins                      | CLUES2 `--df`         |
| `Importance sampling of branch lengths:`  | number of trees to sample from *Relate*              | Relate `--num_samples` |

None of the above prompts can be left blank.

### Optional files (please see CLUES2 and Relate manuals):

`AncientSamps â€” table of sample ages`
`AncientHaps â€” ancient haplotypes`
`Press Enter to skip either file`

Please see the CLUES2 manual for examples of formatting

### Optional parameters to include (please see CLUES2 and Relate manuals):

| prompt                                  | effect                                                              |
| --------------------------------------- | ------------------------------------------------------------------- |
| `Disable allele trajectory? (y/N):`     | answer **y** to *skip* posterior trajectories speeds up inference   |
| `Confidence intervals (e.g. 0.95):`     | compute CI for the selection coefficient (MLE); **Enter** = no CI   |
| `TimeBins (e.g., 200 300):`             | split 0-*tCutoff* into custom intervals; *Enter* = single epoch     |


The example TimeBins 200 300 for a tCutoff 500 yields three epochs: 0-200, 200-300, 300-500.

### Internal housekeeping:

NOTE: All output files will be saved to `~/Output_C2Companion/phase2/{prefix}_chr{N}/`. In this example output files are saved in `~/Output_C2Companion/phase2/FIN_MCM6_original_chr2/)`

### The resulting plot will contain:

SNPs analyzed in Phase 2 \
Confidence intervals \ 
Color intensity bar `[âˆ’logâ‚â‚€(p)]` \ 
Astericks indicating significance above each SNP (* (0.05), ** (0.01), *** (0.001)) 


### A *.tsv file containing SNP information, including related statistics calculated by CLUES2. 

**Example of `*.tsv file`**
```
rsID	POS	der_freq	logLR	-log10(p)	Epoch1_start	Epoch1_end	SelectionMLE1	95%_lower	95%_upper
rs55809728	135842606	0.0909	3.5508	2.11	0	536	0.09981	0.05336	0.14626
rs3754686	135845706	0.6515	5.3420	2.97	0	536	0.02312	0.00405	0.04218
rs4988243	135850133	0.1263	2.0500	1.37	0	536	0.09984	-0.01286	0.21254
rs4954490	135850661	0.6515	5.5070	3.04	0	536	0.02308	0.00369	0.04247
rs4988235	135851076	0.5909	17.2338	8.36	0	536	0.09986	0.08678	0.11294
rs4954493	135852405	0.6515	5.7506	3.16	0	536	0.02420	0.00069	0.04772
rs4988226	135853028	0.6515	5.3420	2.97	0	536	0.02312	0.00405	0.04218
rs309178	135854054	0.6515	5.5244	3.05	0	536	0.02288	0.00368	0.04208
rs309179	135856210	0.6515	5.4909	3.04	0	536	0.02248	0.00139	0.04356
```
[INFO] In addition to the final outputs, users will find three additional folders (`{prefix_inference_chr{N}/`, `prefix_times_chr{N}/`, `prefix_trees_chr{N}/` in which the output from inference.py, RelateToClues.py, and Relate are stored, respectively)

**FuAdditional reading**

CLUES2 manual: https://github.com/avaughn271/CLUES2#command-line-arguments \

Relate SampleBranchLengths: https://myersgroup.github.io/relate/modules.html#SampleBranchLengths \

---

<a name="Phase3"></a>


## Phase 3 â€“ Dating selective sweeps

1 - Apply inference.py (CLUES2) to estimate selection coefficients and confidence intervals in indipendt multi-epochs bin \
2 - Generate Initial onset and writing result into a *.json file \
3 - (optional) Generate bootstraps to estimate confidence interval around first initial onset \
4 - Writing results into a *.json file with summary statistics \

The dating algorithm is based on inferring a selection coefficient in multiple epochs starting from the present and going back into the past. A selection coefficient is computed independently in each window. The assumption for dating is the observation of a selection coefficient that starts deviating from zero (s > 0) and maintains this trend for at least for four younger window. If the first criterion is not satisfied, the script searches for three consecutive windows in which s remains > 0. If there are not even 3 consecutive windows the script searches for 2 consecutive windows but will generate a warning message.


---

### Esample of usage for Phase 3

```
******  CLUES2Companion â€“ please cite CLUES2 and CLUES2Companion  ******
Choose phase to run
  1) Phase 1  : Apply Relate (*.mut, *.anc, *.coal files along with derived allele frequency) 
  2) Phase 2  : Apply Relate (BranchLengths) and CLUES2 (RelateToCLUES.py and inference.py) 
  3) Phase 3  : Date onset of selective sweeps of target SNP(s)
Enter option (1/2/3): 3

```

`then`

```
          â•”â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•—
          â•‘        â±ï¸  PHASE-3 â€“ Dating a selective sweep  â±ï¸         â•‘
          â•‘   Please read the manual carefully before proceeding    â•‘
          â•šâ•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•

Choose chromosome to analyze (e.g. 2, 17, X): 2
Same prefix population name as used in Phase-2 (e.g. Finnish): Finnish 
rsID of SNP to date   (e.g. rs123): rs49988235
df score for CLUES2 (e.g. 600): 600
Initial epoch to scan for initial onset (e.g. 0 or 50): 0
Final epoch to scan for initial onset (e.g. 500 or 1000): 1000
Non overlapping windows size  [default is 50]: 50
```

## Explanation of above command lines

### Script prompts:
Select the chromosome to analyze:
**Chromosome to analyze (e.g., 2, 17, X):**
If you run Phase 1 and Phase 2 on several chromosomes, the pipeline automatically searches for the Phase 1 and Phase 2 outputs that matches the chromosome that you enter above.

**Enter the population prefix:**
Provide the same prefix you used in Phase 1 and Phase 2 (e.g. Finnish). Must exactly match the Phaseâ€‘1/2 prefix; caseâ€‘sensitive. The Phase 3 script now knows the exact folder structure (e.g., output_C2Companion/phase1/Finnish_chr2/).

**Enter the variant identificator (rs) to date:**
Provide the `rs` variant to date (e.g. rs4988235). Only one variant per run. The ID must be present in the phased VCF fed to Phaseâ€‘1.

**Enter the df score (e.g, 600):**
Provide the `df` (--df in CLUES2) score (number of allele frequency bins). we suggest to use the same value employed in Phaseâ€‘2 for comparability.

**Enter the initial epoch to scn for the initial onset:**
Provide the `first` epoch to scan (e.g, 0 or 50)

**Enter the final epoch to scn for the initial onset:**
Provide the `final` epoch to scan (e.g, 500 or 1000). Make sure Final â‰¥ putative sweep age. When unknown, 0â€“1000 is a safe default (â‰ˆâ€¯0â€“28â€¯kya). Otherwise a final epoch of 2000 covers most possible scenarios

**Enter the non overlapping windows size:**
Provide a value representing the window size of two consecutive epochs (e.g, 25 or 50). Ff, as in the example, a starting epoch of 0 and a final (--tcutoff) of 500 and a windows size of 50 are provided, the script will automatically create: 0-50, 50-100, 100-150, 150-200 ..... 450-500 epochs to scan to estimate the selection coefficient in each window indipendentely. Smaller windows means higher computational time cost but less noisy  estimates.

```
e.g, :
Break-points : 0
50
100
150
200
250
300
350
400
450
500
550
600
650
700
750
800
850
900
950
1000
Window size  : 50
```

when the first dating is complete, a message with the summary results will be printed to the terminal alongside a *.json file with the first dating results:

```
e.g., on terminal:
Initial onset â‰ˆ 300 generations (â‰ˆ 8400 years)
epoch : 300 â€“ 350
s(MLE) : 0.02187
```
```
e.g, on *.json:
{
  "rsID": "rs4988235",
  "population": "FIN_MCM63",
  "chromosome": "2",
  "onset_gen": 300,
  "onset_years": 8400,
  "epoch_start": 300,
  "epoch_end": 350,
  "epoch_start_years": 8400,
  "epoch_end_years": 9800,
  "s_MLE": 0.02187
}
```

`After the point estimate the script asks if the users want to proceed with the boostraps:`

```
Proceed with bootstrap dating? [y/N]: y
```
`If the users type "y":`

```
â–¶  Bootstrap settings
Epochs to scan before initial onset (e.g. 200): 50
Epochs to scan after initial onset (e.g. 500): 500
Non overlapping windows size  [default is 25]: 25
Number of bootstrap replicates  [default is 100]: 100
df score for CLUES2  [default is 450]: 600
Importance sampling of branch lengths: 200

Break-points : 50
75
100
125
150
175
200
225
250
275
300
325
350
375
400
425
450
475
500
tCutoff      : 525
Replicates   : 100
```
**[INFO]**: Runtime note.  100 bootstraps with default parameters typically require 16â€“32Â CPUâ€‘hours per SNP.

For the boostraps replicate, the user has to provide again the initial and final epochs to scan. We recommend, in order to save computational time, to use a start and end of the epochs to scan not too far from the first onset. As shown in these examples above, the onset based on the first dating was 300 generations ago. Therefore, epochs for the bootraps 0 - 500 were selected, and to have a more fine distribution, a window size of 25 generations was set. 

The bootraps phase, will create using Relate sample brench lengths, new *.newick files for each replica, then RelateToClues.py will be applied to generate a *_times.txt file for each *.newick file replicated and inference.py (CLUES2) will estimate a selection coefficient in each epoch for each replicates. 
Random seeds (--seed $RANDOM) are always used to generate always different coalescence-based genealogies.

```
e.g.,
â–¶  Generating 100 bootstrap trees & CLUES runs
â†’ [bootstrap 1]  SampleBranchLengths (seed=973675772)
â†’   RelateToCLUES
â†’     CLUES inference â€¢ window [50 - 75]
â†’     CLUES inference â€¢ window [75 - 100]
â†’     CLUES inference â€¢ window [100 - 125]
â†’     CLUES inference â€¢ window [125 - 150]
â†’     CLUES inference â€¢ window [150 - 175]
â†’     CLUES inference â€¢ window [175 - 200]
â†’     CLUES inference â€¢ window [200 - 225]
â†’     CLUES inference â€¢ window [225 - 250]
â†’     CLUES inference â€¢ window [250 - 275]
â†’     CLUES inference â€¢ window [275 - 300]
â†’     CLUES inference â€¢ window [300 - 325]
â†’     CLUES inference â€¢ window [325 - 350]
â†’     CLUES inference â€¢ window [350 - 375]
â†’     CLUES inference â€¢ window [375 - 400]
â†’     CLUES inference â€¢ window [400 - 425]
â†’     CLUES inference â€¢ window [425 - 450]
â†’     CLUES inference â€¢ window [450 - 475]
â†’     CLUES inference â€¢ window [475 - 500]
```

Moreover, the user must decide the number of boostraps, considering the computational times it involves (e.g., 100 bootraps can take from 16 to 32 hours per SNP).

In addition to this, the user must also decide, for the SampleBrenchLenghts.sh step (Relate) the number of trees to sample (importance sampling). For the `df` score and the `importance sampling`, we suggest the user to use the same parameters used in phase2 and phase3.

For the boostraps replicate, the user has to provide again the initial and final epochs to scan. We recommend, in order to save computational time, to use a start and end of the epochs to scan not too far from the first onset. As shown in these examples above, the onset based on the first dating was 300 generations ago. Therefore, epochs for the bootraps 0 - 500 were selected, and to have a more fine distribution, a window size of 25 generations was set. 

At the end of the bootstraps phase, the onset inferred by CLUES2 for each bootraps replication is printed to the terminal. these values â€‹â€‹are used to calculate the summary statistics alongside the *.json.

```
e.g, on *.json:
{
  "rsID": "rs4988235",
  "population": "Finnish_MCM6",
  "chromosome": "2",
  "CI95_low_gen": 155,
  "CI95_high_gen": 344,
  "CI95_low_year": 4340,
  "CI95_high_year": 9632,
  "method": "4 consecutive windows with s > 0"
}
```

<a name="license"></a>

License

MIT License

<a name="contact"></a>

Contact

Alessandro Lisiâ€‚alisi@usc.eduâ€‚(alisi1989)
Michael C. Campbellâ€‚mc44680@usc.eduâ€‚(mc44680)
