<!DOCTYPE html>
<html lang="en">


  <head>

    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-115753061-1"></script>
    <script>
window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());

gtag('config', 'UA-115753061-1');
    </script>


    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title>Relate Documentation</title>
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css">
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>
    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"></script>
    <!-- Scrollbar Custom CSS -->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/malihu-custom-scrollbar-plugin/3.1.5/jquery.mCustomScrollbar.min.css">
    <link rel="stylesheet" href="documentation.css"> 
    <script src="anchor.js"></script>

  </head>


  <body>

    <nav id="myNavbar" class="navbar navbar-default navbar-inverse navbar-fixed-top" role="navigation">
      <!-- Brand and toggle get grouped for better mobile display -->
      <div class="container-fluid">
        <div class="navbar-header">
          <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#navbarCollapse">
            <span class="sr-only">Toggle navigation</span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
          </button>
          <a class="navbar-brand" href="index.html">Relate</a>
        </div>
        <!-- Collect the nav links, forms, and other content for toggling -->
        <div class="collapse navbar-collapse" id="navbarCollapse">
          <ul class="nav navbar-nav">
            <li><a href="index.html">Home</a></li>
            <li><a href="getting_started.html">Getting Started</a></li>
            <li><a href="input_data.html">Input data</a></li>
            <li><a href="modules.html">Add-on modules</a></li>
            <li class="active"><a href="parallelise.html">Parallelise Relate</a></li>
          </ul>
        </div>
      </div>

    </nav>


    <div class="wrapper">

      <!-- Sidebar Holder -->
      <nav id="sidebar">

        <div class="sidebar-header">
          <h3>Overview</h3>
        </div>

        <ul class="list-unstyled components">
          <li>
            <a href="#RelateParallel">Parallelise Relate</a>
          </li>
          <li>
            <a href="#RelateModes">Stages of Relate</a>
          </li>
        </ul>

      </nav>

      <!-- Page Content Holder -->
      <div id="content">




        <div class="container-fluid">

          <!-- row -->
          <div class="row">
            <div class="col-md-10 col-md-offset-1">
              <h3 id="RelateParallel">Parallelise Relate</h3>
              <hr/>

              <p>                 
              <b>Relate</b> has been coded such that it can be parallelised on hundreds of cores on large computing clusters. We provide a script for running Relate on a SGE cluster <a href="parallelise.html#RelateSGEScript">here</a>. For smaller datasets, where it is sufficient to parallelise over only a few cores, we provide a bash script that parallelises <b>Relate</b> without having to submit many jobs.           
              </p>

            </div>
          </div>

          <!-- row -->
          <div class="row">

            <div class="col-md-10 col-md-offset-1">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4 id="RelatePrallelScript">Parallelised Relate</h4></div>
                <div class="panel-body">  

                  <p> 
                  We provide a bash script that parallelises <b>Relate</b> on multiple cores. The script can be found under 
                  <pre><code>PATH_TO_RELATE/scripts/RelateParallel/RelateParallel.sh</code></pre>
                  The script takes the <b>same arguments (and same optional arguments) as Relate</b>, and one additional argument:

                  <dl class="dl-horizontal">                   
                    <dt>--threads</dt>
                    <dd>Maximum number of threads. Default is 1. Note that the script only parallelises <b>Relate</b> if the input data set is sufficiently large. Otherwise, it is more efficient to run <b>Relate</b> on a single core, and <b>Relate</b> will automatically do so. If you set --max_memory to a smaller value, <b>Relate</b> will tend to parallelise over more cores.
                    </dd>
                  </dl>
                  Do not move the script - we are extracting the path to the binaries from the location of the script!
                  </p> 
                  <pre><code> PATH_TO_RELATE/scripts/RelateParallel/RelateParallel.sh\
                  -m 1.25e-8 \
                  -N 30000 \
                  --haps data/example.haps \
                  --sample data/example.sample \
                  --map data/genetic_map.txt \
                  --annot data/example.annot \
                  --seed 1 \
                  -o example \
                  --threads 8 </code></pre>
                  Output: example.anc, example.mut<br/><br/>


                </div>
              </div>

            </div>

          </div>


          <!-- row -->
          <div class="row">

            <div class="col-md-10 col-md-offset-1">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4 id="RelateSGEScript">Relate on a SGE cluster</h4></div>
                <div class="panel-body">  

                  <p> 
                  We provide a bash script for running <b>Relate</b> on a Sun Grid Engine (SGE) cluster. The script can be found under 
                  <pre><code>PATH_TO_RELATE/scripts/RelateSGE/RelateSGE.sh</code></pre>
                  The script takes the <b>same arguments (and same optional arguments) as Relate</b>, and some additional arguments that can be directly mapped to the corresponding SGE arguments:

                  <dl class="dl-horizontal">                   
                    <dt>-P</dt>
                    <dd>Assign job to the project.</dd>
                    <dt>-q</dt>
                    <dd>Queue.</dd>
                    <dt>--pe</dt>
                    <dd>Optional. Processor count per node. Default is 'shmem 1'.</dd>
                  </dl>
                  Do not move the script - we are extracting the path to the binaries from the location of the script!
                  </p> 
                  <pre><code> PATH_TO_RELATE/scripts/RelateSGE/RelateSGE.sh\
                  -m 1.25e-8 \
                  -N 30000 \
                  --haps data/example.haps \
                  --sample data/example.sample \
                  --map data/genetic_map.txt \
                  --annot data/example.annot \
                  --seed 1 \
                  -o example \
                  -P myers.prjc \
                  -q short.qc</code></pre>
                  Output: example.anc, example.mut<br/><br/>


                </div>
              </div>

            </div>

          </div>


          <!-- row -->
          <div class="row">
            <div class="col-md-10 col-md-offset-1">
              <h3 id="RelateModes">Stages of Relate</h3>
              <hr/>
            </div>
          </div>


          <!-- row -->
          <div class="row">

            <div class="col-md-10 col-md-offset-1">

              We implemented the algorithm such that it can be parallelised in multiple ways and, importantly, we can parallelise by submitting hundreds of independent jobs on a computing cluster.<br/>
              To parallelise <b>Relate</b>, do not use mode All. Instead we have to execute each of the 7 stages in sequence: <br/>

              <ol>
                <li>MakeChunks: Divides the chromosome into chunks. To each chunk, we apply stages 2-6.</li>
                <li>Paint: Applies the chromosome painting algorithm and stores distance matrices as temporary files.</li>
                <li>BuildTopology: Uses distance matrices calculated in the previous stage and estimates tree topologies.</li>
                <li>FindEquivalentBranches: Looks for equivalent branches in adjacent trees.</li>
                <li>InferBranchLengths: Estimates branch lengths for the trees.</li>
                <li>CombineSections: Combines files for the chunk.</li>
                <li>Finalize: Combines all chunks and generates the output.</li>
              </ol>

              Let us define the following parameters:
              <ul>
                <li>output [string]: Filename of output without file extension.</li>
                <li>mu [dbl]: mutation rate.</li>
                <li>Ne [dbl]: effective population size of haplotypes. (NOT of individuals! To get the population size of haplotypes, multiply the effective population size of individuals by 2.)</li>
              </ul>

            </div>

            <div class="col-md-10 col-md-offset-1">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4 id="MakeChunks">1. Make chunks</h4></div>
                <div class="panel-body">  

                  <p> 
                  <ul>
                    <li>Argument --haps, --sample, and --map take the filenames of the input data.</li>
                    <li>Optional argument --dist takes the filename of a .dist file. A .dist file specifies distances (in BP) between SNPs. This is to correct for regions with low mappability in the genomic mask (A tool that takes a genomic mask and generates a .dist file can be found <a href="input_data.html#FilterHapsUsingMask">here</a>).</li>
                    <li>Optional argument --memory option specifies the approximate memory allowance (in GB) used for storing the distance matrices in memory. <b>Relate</b> will exceed this amount, as it also stores trees and other information in memory; however the option allows to approximately control memory usage. Default is 5GB. <b>Relate</b> can become more efficient (in runtime and hard disk usage) with more memory, particularly for large sample sizes. </li>
                  </ul>

                  </p> 
                  <pre><code> PATH_TO_RELATE/bin/Relate \
                  --mode "MakeChunks" \
                  --haps data/example.haps \
                  --sample data/example.sample \
                  --map data/genetic_map.txt \
                  -o ${output}</code></pre>
                  Output: temporary files<br/>

                </div>
              </div>

            </div>

          </div>


          <!-- row -->
          <div class="row">

            <div class="col-md-10 col-md-offset-1">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4 id="for_loop1">For loop</h4></div>
                <div class="panel-body">  

                  First, obtain the number of chunks created in the previous step. In bash, this can be done as follows
                  <p> 
                  <pre><code>num_chunks=$(ls parameters_* | wc -l)</code></pre>
                  </p>

                  <p>
                  Then, apply steps 2 - 6 to each chunk
                  <pre><code> for chunk_index in `seq 0 $(($num_chunks - 1))`;
                  do
                    #stages 2-6
                  done</code></pre>
                  </p>

                  <p>
                  <b>This for loop can be parallelised</b> without moving any files or changing directories.
                  </p>


                </div>
              </div>

            </div>

          </div>


          <!-- row -->
          <div class="row">

            <div class="col-md-10 col-md-offset-1">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4 id="Paint">2. Paint</h4></div>
                <div class="panel-body">  

                  <p> 
                  <ul>
                    <li>Argument --chunk_index specifies the index [int] of a chunk.</li>
                  </ul>

                  </p> 
                  <pre><code> PATH_TO_RELATE/bin/Relate \
                  --mode Paint \
                  --chunk_index ${chunk_index} \
                  -o ${output}</code></pre>
                  Output: temporary files<br/>

                </div>
              </div>

            </div>

          </div>

          <!-- row -->
          <div class="row">

            <div class="col-md-10 col-md-offset-1">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4 id="BuildTopolgy">3. Build Topology</h4></div>
                <div class="panel-body">  

                  <p> 
                  <ul>
                    <li>Argument --chunk_index specifies the index [int] of a chunk.</li>
                    <li>Optional arguments --first_section, --last_section specify the first and last section within the chunk, for which to infer tree topologies. With these options, we can parallelise this mode (see below).</li>
                    <li>Argument -o specifies the output filename without file extension.</li>
                  </ul>

                  </p> 
                  <pre><code> PATH_TO_RELATE/bin/Relate \
                  --mode BuildTopology \
                  --chunk_index ${chunk_index} \
                  -o ${output} </code></pre>
                  Output: temporary files<br/><br/>

                  <p>
                  <b>Optional:</b><br/>
                  You can code another parallelisable for loop here. To do this, you need to specify options first_section and last_section. First obtain the number of sections for this chunk. In bash, this can be done as follows
                  <p> 
                  <pre><code>num_sections=$(ls chunk_${chunk_index}/paint/*bin | wc -l)</code></pre>
                  </p>

                  You can now code a for loop,
                  <pre><code> for section in `seq 0 $(($num_sections - 1))`;
                  do
                    PATH_TO_RELATE/bin/Relate \
                        --mode BuildTopology \
                        --first_section $section \
                        --last_section $section \
                        --chunk_index ${chunk_index} \
                        -o ${output} 
                  done</code></pre>
                  Here --first_section and --last_section are both set to $section, such that only trees in one section will be inferred at a time. You can also specify more than one section here.
                  </p>


                  <p>
                  <b>This for loop can be parallelised</b> without moving any files or changing directories.
                  </p>

                </div>
              </div>

            </div>

          </div>

          <!-- row -->
          <div class="row">

            <div class="col-md-10 col-md-offset-1">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4 id="FindEquivalentBranches">4. Find equivalent branches in adjacent trees and propagate mutations</h4></div>
                <div class="panel-body">  

                  <p> 
                  <ul>
                    <li>Argument --chunk_index specifies the index [int] of a chunk.</li>
                    <li>Argument -o specifies the output filename without file extension.</li>
                  </ul>

                  </p> 
                  <pre><code> PATH_TO_RELATE/bin/Relate \
                  --mode "FindEquivalentBranches" \
                  --chunk_index ${chunk_index} \
                  -o ${output} </code></pre>
                  Output: temporary files<br/>

                </div>
              </div>

            </div>

          </div>

          <!-- row -->
          <div class="row">

            <div class="col-md-10 col-md-offset-1">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4 id="InferBranchLengths">5. Estimate branch lengths</h4></div>
                <div class="panel-body">  

                  <p> 
                  <ul>
                    <li>Argument -m specifies the mutation rate [dbl]</li>
                    <li>Argument -N specifies the effective population size [dbl]</li>
                    <li>Argument --chunk_index specifies the index ([integer]) of a chunk.</li>
                    <li>Optional arguments --first_section, --last_section specify the first and last section within the chunk, for which to infer branch lengths. With these options, we can parallelise this mode (similar to <a href="#BuildTopology">stage 3</a> of the algorithm).</li>
                    <li>Argument -o specifies the output filename without file extension.</li>
                  </ul>

                  </p> 
                  <pre><code> PATH_TO_RELATE/bin/Relate \
                  --mode "InferBranchLengths" \
                  -m $mu \
                  -N $Ne \
                  --chunk_index ${chunk_index} \
                  -o ${output} </code></pre>
                  Output: temporary files<br/><br/>

                  <b>Optional:</b><br/>
                  Similarly to <a href="#BuildTopology">stage 3</a>, you can code a for loop,
                  <pre><code> for section in `seq 0 $(($num_sections - 1))`;
                  do
                    PATH_TO_RELATE/bin/Relate \
                        --mode "InferBranchLengths" \
                        -m $mu \
                        -N $Ne \
                        --first_section $section \
                        --last_section $section \
                        --chunk_index ${chunk_index} \
                        -o ${output} 
                  done</code></pre>
                  Here --first_section and --last_section are both set to $section, such that only trees in one section will be inferred at a time. You can also specify more than one section here. Optionally, you can specify the --coal option, with which we specify coalescence rates through time in form of a .coal file. See <a href="modules.html#.coal"> here </a> for file format. Specifying this option will overwrite the --effectiveN option.
                  </p>

                  <p>
                  <b>This for loop can be parallelised</b> without moving any files or changing directories.
                  </p>



                </div>
              </div>

            </div>

          </div>

          <!-- row -->
          <div class="row">

            <div class="col-md-10 col-md-offset-1">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4 id="CombineArgs">6. Combine files</h4></div>
                <div class="panel-body">  

                  <p> 
                  <ul>
                    <li>Argument -N specifies the effective population size [dbl]</li>
                    <li>Argument --chunk_index specifies the index [int] of a chunk.</li>
                    <li>Argument -o specifies the output filename without file extension.</li>

                  </ul>

                  </p> 
                  <pre><code> PATH_TO_RELATE/bin/Relate \
                  --mode "CombineSections" \
                  -N $Ne \
                  --chunk_index ${chunk_index} \
                  -o ${output} </code></pre>
                  Output: temporary files<br/>

                </div>
              </div>

            </div>

          </div>

          <!-- row -->
          <div class="row">

            <div class="col-md-10 col-md-offset-1">
              <div class="panel panel-info">
                <div class="panel-heading"> <h4 id="Finalize">7. Finalize output</h4></div>
                <div class="panel-body">  

                  <p> 
                  <ul>
                    <li>Argument -o specifies the output filename without file extension.</li>
                  </ul>

                  </p> 
                  <pre><code> PATH_TO_RELATE/bin/Relate \
                  --mode "Finalize" \
                  -o ${output} </code></pre>
                  Output: ${output}.anc, ${output}.mut<br/>

                </div>
              </div>

            </div>

          </div>

        </div>

      </div>

    </div>

    <script>
      anchors.add();  
    </script>

  </body>
</html>     
